File: java/src/main/java/br/com/emendes/problems/AddBinary.java
package br.com.emendes.problems;

/**
 * Dado duas strings binários a e b, retorne a soma como uma string.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= a.length, b.length <= 10⁴<br>
 * a e b consistem apenas de caracteres 0 e 1.<br>
 * Cada string não contem 0 iniciais, exceto o próprio valor 0.
 */
public class AddBinary {

  public String addBinary(String a, String b) {
    int aLength = a.length();
    int bLength = b.length();
    int iA = aLength - 1;
    int iB = bLength - 1;
    StringBuilder answer = new StringBuilder();
    int carrier = 0;
    if (aLength > bLength) {
      for (; iB >= 0; iB--, iA--) {
        int sum = (a.charAt(iA) - 48) + (b.charAt(iB) - 48) + carrier;
        answer.append(sum % 2);
        carrier = sum / 2;
      }
      for (; iA >= 0; iA--) {
        int sum = (a.charAt(iA) - 48) + carrier;
        answer.append(sum % 2);
        carrier = sum / 2;
      }
    } else {
      for (; iA >= 0; iA--, iB--) {
        int sum = (a.charAt(iA) - 48) + (b.charAt(iB) - 48) + carrier;
        answer.append(sum % 2);
        carrier = sum / 2;
      }
      for (; iB >= 0; iB--) {
        int sum = (b.charAt(iB) - 48) + carrier;
        answer.append(sum % 2);
        carrier = sum / 2;
      }
    }
    if (carrier == 1) {
      answer.append(1);
    }
    answer.reverse();

    return answer.toString();
  }

}

File: javascript/problems/AddBinary.js
/**
 * Dadas duas strings binárias a e b, retorne sua soma como uma string binária.
 *
 * Restrições:
 *
 * 1 <= a.length, b.length <= 10⁴
 *
 * a e b consistem apenas de caracteres '0' ou '1'.
 *
 * Cada string não contém zeros a esquerda exceto o próprio zero.
 *
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
const addBinary = function (a, b) {
  const result = [];
  let carry = 0;
  let i = a.length - 1;
  let j = b.length - 1;
  while (i >= 0 || j >= 0 || carry === 1) {
    if (i >= 0) carry += +a[i--];
    if (j >= 0) carry += +b[j--];
    result.unshift(carry % 2);
    carry = carry > 1 ? 1 : 0;
  }
  return result.join("");
};
